src/libre/ast_compile_internal.h

```diff
@@ -0,0 +1,162 @@
+
+	LINK_NONE = 0x00
+};
+
+/* Call stack for AST -> NFA conversion.
+ *
+ * This is executed depth first and could otherwise use the C call
+ * stack, but it needs to support forking execution in order to handle
+ * AST_EXPR_ALTs where ALT subtrees have different capture behavior.
+ * Passing through AST_EXPR_GROUP nodes affect the remainder of
+ * evaluation (even after returning from their subtree), because they
+ * modify which captures will be associated with particular end states.
+ * When they only appear in some ALTs we need to fork the remainder of
+ * execution, including evaluating everything pending on the call stack
+ * multiple times. */
+#define DEF_COMP_STACK_CEIL 4
+struct comp_stack {
+	size_t ceil;
+	size_t depth;		/* 0 -> empty */
+
+	fsm_state_t end_inner;
+	fsm_state_t end_outer;
+	fsm_state_t end_any_loop;
+	fsm_state_t end_any_inner;
+	int have_end_any_loop;
+
+	struct comp_stack_frame {
+		struct ast_expr *n;
+		fsm_state_t x;
+		fsm_state_t y;
+		unsigned step;
+
+		union {
+			struct {
+				fsm_state_t link;
+			} concat;
+			struct {
+				unsigned count;
+			} alt;
+			struct {
+				struct fsm_subgraph subgraph;
+				fsm_state_t na;
+				fsm_state_t nz;
+			} repeat;
+		} u;
+	} *frames;
+};
+
+struct comp_env {
+	const struct fsm_alloc *alloc;
+	struct fsm *fsm;
+	enum re_flags re_flags;
+	struct re_err *err;
+
+	/*
+	 * These are saved so that dialects without implicit
+	 * anchoring can create states with '.' edges to self
+	 * on demand, and link them to the original start and
+	 * end states.
+	 *
+	 * Also, some states in a first/last context need to link
+	 * directly to the overall start/end states, either in
+	 * place of or along with the adjacent states.
+	 *
+	 * The inner start and end states are considered inside of
+	 * match group 0, outer are not.
+	 */
+	fsm_state_t start_outer;
+	fsm_state_t start_inner;
+	fsm_state_t start_any_loop;
+	fsm_state_t start_any_inner;
+	int have_start_any_loop;
+};
+
+static int
+comp_iter_trampoline(struct comp_env *env,
+	fsm_state_t x, struct ast_expr *n);
+
+static int
+eval_stack_frame(struct comp_env *env, struct comp_stack *stack);
+
+static int
+eval_EMPTY(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_CONCAT(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_ALT(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_LITERAL(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_CODEPOINT(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_REPEAT(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_GROUP(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_ANCHOR(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_SUBTRACT(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_RANGE(struct comp_env *env, struct comp_stack *stack);
+static int
+eval_TOMBSTONE(struct comp_env *env, struct comp_stack *stack);
```

Comment:
i wouldn't include static prototypes unless you need them (e.g. for mutually recursive functions)

Reply:
These are forward references.